### PR TITLE
Channel uri receiver window

### PR DESCRIPTION
**Pull request by mikeb01**
- Adds the ability to set the receiver initial window in the URI.
- Adds validation to C and Java media drivers that compare the mtu to the socket sndbuf and receiver window to the socket rcvbuf.
- Adds validation to prevent creating subscriptions or publications that try to change the socket sndbuf and rcvbuf values.

### CODE REVIEW

File: aeron-client/src/main/java/io/aeron/ChannelUriStringBuilder.java [hunk 1]
```diff
@@ -70,6 +70,7 @@
     private boolean isSessionIdTagged;
     private Integer socketSndbufLength;
     private Integer socketRcvbufLength;
+    private Integer receiverWindowLength;
```
Comment (mjpt777) [inline aeron-client/src/main/java/io/aeron/ChannelUriStringBuilder.java:73]: The reset of this value is missing from the `clear()` method.

File: aeron-driver/src/main/c/media/aeron_receive_channel_endpoint.h [hunk 2]
```diff
@@ -186,7 +188,10 @@ int aeron_receive_channel_endpoint_on_remove_publication_image(
     aeron_receive_channel_endpoint_t *endpoint, aeron_publication_image_t *image);
 
 int aeron_receiver_channel_endpoint_validate_sender_mtu_length(
-    aeron_receive_channel_endpoint_t *endpoint, size_t sender_mtu_length, size_t window_max_length);
+    aeron_receive_channel_endpoint_t *endpoint,
+    size_t sender_mtu_length,
+    size_t window_max_length,
+    aeron_driver_context_t *pStct);
```
Comment (mjpt777) [inline aeron-driver/src/main/c/media/aeron_receive_channel_endpoint.h:194]: Naming convention is not camel case in C.

File: aeron-driver/src/main/java/io/aeron/driver/Configuration.java [hunk 3]
```diff
@@ -1693,7 +1689,8 @@ public static void validateInitialWindowLength(final int initialWindowLength, fi
     {
         if (mtuLength > initialWindowLength)
         {
-            throw new ConfigurationException("initial window length must be >= to MTU length: " + mtuLength);
+            throw new ConfigurationException(
+                "initial window length (" + initialWindowLength + ") must be >= MTU length (" + mtuLength + ")");
```
Comment (mjpt777) [inline aeron-driver/src/main/java/io/aeron/driver/Configuration.java:1693]: `"mtuLength=" + mtuLength + " > initialWindowLength=" + initialWindowLength ` is easier to reason about given check.

File: aeron-driver/src/main/java/io/aeron/driver/Configuration.java [hunk 4]
```diff
@@ -1767,55 +1764,43 @@ public static TerminationValidator terminationValidator()
...
-            if (maxSoRcvBuf < ctx.socketRcvbufLength())
-            {
-                System.err.format(
-                    "WARNING: Could not get desired SO_RCVBUF, adjust OS to allow %s: attempted=%d, actual=%d%n",
-                    SOCKET_RCVBUF_LENGTH_PROP_NAME,
-                    ctx.socketRcvbufLength(),
-                    maxSoRcvBuf);
-            }
+            System.err.format(
+                "WARNING: Could not get desired SO_SNDBUF, adjust OS to allow %s: attempted=%d, actual=%d%n",
```
Comment (mjpt777) [inline aeron-driver/src/main/java/io/aeron/driver/Configuration.java:1770]: get -> set

File: aeron-driver/src/main/java/io/aeron/driver/Configuration.java [hunk 5]
```diff
@@ -1767,55 +1764,43 @@ public static TerminationValidator terminationValidator()
...
 
-            if (ctx.initialWindowLength() > maxSoRcvBuf)
-            {
-                throw new ConfigurationException("window length greater than socket SO_RCVBUF, increase '" +
-                    Configuration.INITIAL_WINDOW_LENGTH_PROP_NAME +
-                    "' to match window: windowLength=" + ctx.initialWindowLength() + ", SO_RCVBUF=" + maxSoRcvBuf);
-            }
+        if (ctx.mtuLength() > soSndBuf)
+        {
+            throw new ConfigurationException(String.format(
```
Comment (mjpt777) [inline aeron-driver/src/main/java/io/aeron/driver/Configuration.java:1790]: By default we should avoid use of string format.

**Comment by mjpt777** [comment]
This is a nice step forward.

File: aeron-driver/src/main/java/io/aeron/driver/Configuration.java (hunk 3 above)
Comment (mikeb01) [reply aeron-driver/src/main/java/io/aeron/driver/Configuration.java:1693]: I've changed to this, but I've modified the C code to have the same error format as it simplifies the error checking in the system tests and results in a more consistent experience for users.  However in the C driver the naming convention won't match.

**Pull request merged**